### DISCUSSION TITLE
Adopting langchain generations format

### DISCUSSION THREAD

**Discussion by ghost** [BerriAI/litellm#98] [Ideas]
Wondering if it might make more sense to adopt something similar to the langchain Generations format

- LiteLLM output
- original model provider response 

Reason I like this is because it gives me an easy way to tell if something went wrong - who was at fault (could maybe also include the exact input params to the model, for easy repro). 

<img width="530" alt="Screenshot 2023-08-11 at 8 59 50 AM" src="https://github.com/BerriAI/litellm/assets/17561003/4725e79c-b5d8-4032-a98c-9a394e93d6f0">

**Comment by ghost** [comment]
cc: @ishaan-jaff

**Comment by ghost** [comment]
<img width="668" alt="Screenshot 2023-08-11 at 9 06 07 AM" src="https://github.com/BerriAI/litellm/assets/17561003/a080994d-3b23-4d5a-99d9-30595d7389e9">

https://github.com/refuel-ai/autolabel also has an interesting approach here 
- response 
- error (if any)
- cost

**Comment by kinthaiofficial** [comment]
The LangChain generations format is worth adopting as the baseline for LiteLLM's response structure — it gives you a common interface that LangChain-based pipelines can consume directly, which dramatically simplifies integration.

Key things to get right:

**Generation metadata passthrough**
The `generation_info` field in LangChain's `ChatGeneration` should carry through: `finish_reason`, `model_name`, `token_usage` (including `prompt_tokens`, `completion_tokens`, `total_tokens`), and provider-specific extras. Losing any of these forces downstream consumers to maintain their own extraction logic.

**Streaming compatibility**
LangChain's streaming uses `ChatGenerationChunk` with `.text` and `.generation_info`. LiteLLM's streaming response should be compatible with this so that streaming chains work without wrapping.

**Tool call representation**
When the model returns a tool call instead of text, LangChain uses `AIMessage` with `tool_calls` as a list of structured dicts. LiteLLM should normalize provider-specific tool call formats (Anthropic uses `tool_use` content blocks, OpenAI uses a `tool_calls` array) into this single representation.

**Logprobs if available**
Some applications use logprobs for confidence scoring and calibration. LangChain's response format can carry these — LiteLLM should pass them through when the provider returns them.

The value of this standardization goes beyond LangChain compatibility — it means any metrics/logging layer that knows the LangChain format gets LiteLLM observability for free.
